src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -0,0 +1,369 @@
+        '''Perform various checks to ensure that it is valid to apply the
+        ArrayRange2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Assignment`
+
+        :raises TransformationError: if the node argument is not an \
+            Assignment.
+        :raises TransformationError: if the node argument is an \
+            Assignment whose left had side is not an Array.
+        :raises TransformationError: if the node argument is an \
+            Assignment whose left hand side is an Array that does not \
+            have Range specifying the access to at least one of its \
+            dimensions.
+        :raises TransformationError: if two or more of the loop ranges \
+        in the assignment are different or are not known to be the \
+        same.
+
+        '''
+        if not isinstance(node, Assignment):
+            raise TransformationError(
+                "Error in {0} transformation. The supplied node argument "
+                "should be a PSyIR Assignment, but found '{1}'."
+                "".format(self.name, type(node).__name__))
+
+        if not isinstance(node.lhs, Array):
+            raise TransformationError(
+                "Error in {0} transformation. The lhs of the supplied "
+                "Assignment node should be a PSyIR Array, but found '{1}'."
+                "".format(self.name, type(node.lhs).__name__))
+
+        if not [dim for dim in node.lhs.children if isinstance(dim, Range)]:
+            raise TransformationError(
+                "Error in {0} transformation. The lhs of the supplied "
+                "Assignment node should be a PSyIR Array with at least one "
+                "of its dimensions being a Range, but found None."
+                "".format(self.name))
+
+        # If an operator on the rhs only returns an array then we are
+        # not able to turn the assignment into an explicit loop. At
+        # the moment we do not capture what an operator returns and
+        # therefore can not check this. This will be addressed in
+        # issue #685. For the moment we check with an explicit list
+        # which happens to only contain a single operator (MATMUL)
+        # because at this time all other operators in the PSyIR can be
+        # performed elementwise.
+        if [operation for operation in node.rhs.walk(Operation)
+                if operation.operator in [BinaryOperation.Operator.MATMUL]]:
+            raise TransformationError(
+                "Error in {0} transformation. The rhs of the supplied "
+                "Assignment node contains the MATMUL operator which can't be "
+                "performed elementwise.".format(self.name))
+
+        for array in node.walk(Array):
+            for idx, child in enumerate(array.children):
+                if isinstance(child, Range):
+                    if array is node.lhs:
```

Comment:
Rather than having a separate loop/walk for the lhs and rhs I merged the loops/walks together. I could do this as we go depth first in the walk so will always hit the lhs first. Therefore we store that when we see it and it is then available in the else clause for any rhs checks. I've split the loops out to make it clearer. I also realised that this check is for the leftmost ranges and we have changed to modifying the rightmost ranges. Therefore I've updated the check appropriately.